packages/storybook/src/stories/grid-error.stories.tsx

```diff
@@ -0,0 +1,262 @@
+    }
+  }, [api]);
+
+  return (
+    <div className="grid-container">
+      <XGrid rows={rows} columns={cols} apiRef={api} />
+    </div>
+  );
+};
+export const errorProp = () => {
+  const rows = React.useMemo(() => getRows(), []);
+  const cols = React.useMemo(() => getColumns(), []);
+
+  return (
+    <div className="grid-container">
+      <XGrid rows={rows} columns={cols} error={{ message: 'Error can also be set in props!' }} />
+    </div>
+  );
+};
+function CustomErrorOverlay(props) {
+  return (
+    <GridOverlay className="custom-overlay">
+      <div style={{ textAlign: 'center' }}>
+        <h1>{props.title}</h1>
+        <p>{typeof props.error === 'string' ? props.error : props.error.message}</p>
+      </div>
+    </GridOverlay>
+  );
+}
+
+export const CustomError = () => {
+  const api = useApiRef();
+  const rows = React.useMemo(() => getRows(), []);
+  const cols = React.useMemo(() => getColumns(), []);
+
+  React.useEffect(() => {
+    if (api && api.current) {
+      api.current!.showError({ error: 'Something bad happened!', title: 'BIG ERROR' });
+    }
+  }, [api]);
+
+  return (
+    <div className="grid-container">
+      <XGrid
+        rows={rows}
+        columns={cols}
+        apiRef={api}
+        components={{
+          errorOverlay: CustomErrorOverlay,
+        }}
+      />
+    </div>
+  );
+};
+export const CustomErrorWithException = () => {
+  const rows = React.useMemo(() => getRows(), []);
+  const cols = React.useMemo(() => getColumns(), []);
+  cols[1].cellClassRules = {
+    common: ({ value }) => {
+      if (value === 'alice') {
+        throw new Error('Alice created an error!');
+      }
+      return true;
+    },
+  };
+
+  return (
+    <div className="grid-container">
+      <XGrid
+        rows={rows}
+        columns={cols}
+        components={{
+          errorOverlay: CustomErrorOverlay,
+        }}
+      />
+    </div>
+  );
+};
+function fetchError(): Promise<any> {
+  return new Promise<any>((resolve, reject) => {
+    setTimeout(() => {
+      reject(new Error('Bad fetch'));
+    }, 500);
+  });
+}
+
+export const AsyncErrorApi = () => {
+  const api = useApiRef();
+  const rows = React.useMemo(() => getRows(), []);
+  const cols = React.useMemo(() => getColumns(), []);
+
+  React.useEffect(() => {
+    fetchError().catch((err) => {
+      api.current!.showError(err);
+    });
+  }, [api]);
+
+  return (
+    <div className="grid-container">
+      <XGrid rows={rows} columns={cols} apiRef={api} />
+    </div>
+  );
+};
+
+export const onErrorHandler = () => {
+  const rows = React.useMemo(() => getRows(), []);
+  const cols = React.useMemo(() => getColumns(), []);
+  const [errorMessage, setErrorMessage] = React.useState('');
+
+  cols[1].cellClassRules = {
+    common: ({ value }) => {
+      if (value === 'alice') {
+        throw new Error('Alice created an error!');
+      }
+      return true;
+    },
+  };
+
+  const onError = React.useCallback(
```

Comment:
This `useCallback` requirement still makes me tick. I can't help myself, what about this diff?

```diff
diff --git a/packages/grid/x-grid-modules/src/hooks/root/useApiEventHandler.ts b/packages/grid/x-grid-modules/src/hooks/root/useApiEventHandler.ts
index 20bccc98..7353c7b1 100644
--- a/packages/grid/x-grid-modules/src/hooks/root/useApiEventHandler.ts
+++ b/packages/grid/x-grid-modules/src/hooks/root/useApiEventHandler.ts
@@ -1,4 +1,5 @@
 import * as React from 'react';
+import { useEventCallback } from '@material-ui/core/utils';
 import { useLogger } from '../utils';
 import { ApiRef } from '../../models/api';

@@ -9,11 +10,21 @@ export function useApiEventHandler(
 ) {
   const logger = useLogger('useApiEventHandler');

+  const handleEvent = useEventCallback((...args) => {
+    if (handler) {
+      handler(...args);
+    }
+  });
+
   React.useEffect(() => {
-    if (apiRef && apiRef.current && handler && eventName) {
-      return apiRef.current!.registerEvent(eventName, handler);
+    if (!apiRef || !apiRef.current) {
+      throw new Error('Material-UI: Wrong binding of apiRef.');
+    }
+
+    if (eventName) {
+      return apiRef.current!.registerEvent(eventName, handleEvent);
     }

     return undefined;
-  }, [apiRef, logger, eventName, handler]);
+  }, [apiRef, logger, eventName, handleEvent]);
 }
```

Reply:
Well the requirement is to use immutables hence why we use useCallback 
If you use class component you won't need this one

I suggest we start with strong immutability as it is, and see how it goes for our users.

Reply:
Sound fair, let's do that, opening an issue for later.